File: Windows/SD.IdentitySystem.IPresentation/Models/Outputs/Menu.cs
﻿using SD.Infrastructure.PresentationBase;
using SD.Toolkits.Recursion.Tree;
using System;
using System.Collections.Generic;

namespace SD.IdentitySystem.IPresentation.Models.Outputs
{
    /// <summary>
    /// 菜单模型
    /// </summary>
    public class Menu : Model, ITree<Menu>
    {
        #region 构造器
        /// <summary>
        /// 构造器
        /// </summary>
        public Menu()
        {
            //初始化导航属性
            this.SubNodes = new HashSet<Menu>();

            //默认值
            this.IsChecked = false;
        }
        #endregion

        #region 链接地址 —— string Url
        /// <summary>
        /// 链接地址
        /// </summary>
        public string Url { get; set; }
        #endregion

        #region 路径 —— string Path
        /// <summary>
        /// 路径
        /// </summary>
        public string Path { get; set; }
        #endregion

        #region 图标 —— string Icon
        /// <summary>
        /// 图标
        /// </summary>
        public string Icon { get; set; }
        #endregion

        #region 是否是根级节点 —— bool IsRoot
        /// <summary>
        /// 是否是根级节点
        /// </summary>
        public bool IsRoot { get; set; }
        #endregion

        #region 是否是叶子级节点 —— bool IsLeaf
        /// <summary>
        /// 是否是叶子级节点
        /// </summary>

        public bool IsLeaf { get; set; }
        #endregion

        #region 排序 —— int Sort
        /// <summary>
        /// 排序
        /// </summary>
        public int Sort { get; set; }
        #endregion

        #region 上级菜单Id —— Guid? ParentMenuId
        /// <summary>
        /// 上级菜单Id
        /// </summary>
        public Guid? ParentMenuId { get; set; }
        #endregion


        //Others

        #region 信息系统编号 —— string SystemNo
        /// <summary>
        /// 信息系统编号
        /// </summary>
        public string SystemNo { get; set; }
        #endregion

        #region 信息系统名称 —— string SystemName
        /// <summary>
        /// 信息系统名称
        /// </summary>
        public string SystemName { get; set; }
        #endregion

        #region 应用程序类型 —— string ApplicationType
        /// <summary>
        /// 应用程序类型
        /// </summary>
        public string ApplicationType { get; set; }
        #endregion

        #region 导航属性 - 上级菜单 —— ICollection<Menu> ParentNode
        /// <summary>
        /// 导航属性 - 上级菜单
        /// </summary>
        public Menu ParentNode { get; set; }
        #endregion

        #region 导航属性 - 下级菜单集 —— ICollection<Menu> SubNodes
        /// <summary>
        /// 导航属性 - 下级菜单集
        /// </summary>
        public ICollection<Menu> SubNodes { get; set; }
        #endregion
    }
}

File: Angular/SD.IdentitySystem.Client/src/models/menu.ts
import {ApplicationType} from "../enums/applicationType";

/*菜单*/
export class Menu {
    /*信息系统编号*/
    public systemNo: string;

    /*信息系统名称*/
    public systemName: string;

    /*应用程序类型*/
    public applicationType;

    /*菜单Id*/
    public id: string;

    /*菜单名称*/
    public name: string;

    /*链接地址*/
    public url: string;

    /*路径*/
    public path: string;

    /*图标*/
    public icon: string;

    /*排序*/
    public sort: number;

    /*层次*/
    public level: number;

    /*是否是根级节点*/
    public isRoot: boolean;

    /*是否是叶子级节点*/
    public isLeaf: boolean;

    /*上级菜单Id*/
    public parentMenuId: string;

    /*上级菜单*/
    public parentNode: Menu | null;

    /*下级菜单集*/
    public subNodes: Array<Menu>;

    /**
     * 创建菜单构造器
     * */
    public constructor(systemNo: string, systemName: string, applicationType: ApplicationType, menuId: string, menuName: string, url: string, path: string, icon: string, sort: number, level: number, isRoot: boolean, isLeaf: boolean, parentMenuId: string, parentNode: Menu | null) {
        this.systemNo = systemNo;
        this.systemName = systemName;
        this.applicationType = applicationType;
        this.id = menuId;
        this.name = menuName;
        this.url = url;
        this.path = path;
        this.icon = icon;
        this.sort = sort;
        this.level = level;
        this.isRoot = isRoot;
        this.isLeaf = isLeaf;
        this.parentMenuId = parentMenuId;
        this.parentNode = parentNode;
        this.subNodes = new Array<Menu>();
        if (parentNode != null) {
            parentNode.subNodes.push(this);
        }
    }
}
